config/src/main/java/org/springframework/security/config/web/server/ServerHttpSecurity.java

```diff
@@ -1907,6 +1957,234 @@ public AuthorizeExchangeSpec access(ReactiveAuthorizationManager<AuthorizationCo
+		 */
+		public SessionManagementSpec concurrentSessions(Customizer<ConcurrentSessionsSpec> customizer) {
+			if (this.concurrentSession == null) {
+				this.concurrentSession = new ConcurrentSessionsSpec();
+			}
+			customizer.customize(this.concurrentSession);
+			return this;
+		}
+
+		public SessionManagementSpec sessionAuthenticationStrategy(
+				ServerSessionAuthenticationStrategy sessionAuthenticationStrategy) {
+			Assert.notNull(sessionAuthenticationStrategy, "sessionAuthenticationStrategy cannot be null");
+			this.providedSessionAuthenticationStrategy = sessionAuthenticationStrategy;
+			return this;
+		}
+
+		void configure(ServerHttpSecurity http) {
+			if (this.concurrentSession != null) {
+				ReactiveSessionRegistry reactiveSessionRegistry = getSessionRegistry();
+				ConcurrentSessionControlServerSessionAuthenticationStrategy concurrentSessionControlStrategy = new ConcurrentSessionControlServerSessionAuthenticationStrategy(
+						reactiveSessionRegistry);
+				concurrentSessionControlStrategy.setMaximumSessions(this.maxSessions);
+				concurrentSessionControlStrategy.setMaximumSessionsExceededHandler(this.maximumSessionsExceededHandler);
+				RegisterSessionServerSessionAuthenticationStrategy registerSessionAuthenticationStrategy = new RegisterSessionServerSessionAuthenticationStrategy(
+						reactiveSessionRegistry);
+				this.sessionAuthenticationStrategy = new DelegatingServerSessionAuthenticationStrategy(
+						concurrentSessionControlStrategy, this.providedSessionAuthenticationStrategy,
+						registerSessionAuthenticationStrategy);
+				SessionRegistryWebFilter sessionRegistryWebFilter = new SessionRegistryWebFilter(
+						reactiveSessionRegistry);
+				http.addFilterAfter(sessionRegistryWebFilter, SecurityWebFiltersOrder.HTTP_HEADERS_WRITER);
+			}
+			else {
+				this.sessionAuthenticationStrategy = this.providedSessionAuthenticationStrategy;
+			}
+		}
+
+		private ReactiveSessionRegistry getSessionRegistry() {
+			if (this.sessionRegistry == null) {
+				this.sessionRegistry = getBeanOrNull(ReactiveSessionRegistry.class);
+			}
+			if (this.sessionRegistry == null) {
+				throw new IllegalStateException(
+						"A ReactiveSessionRegistry needs to be specified to enable concurrent session management");
+			}
+			return this.sessionRegistry;
+		}
+
+		/**
+		 * Configures how many sessions are allowed for a given user.
+		 */
+		public class ConcurrentSessionsSpec {
+
+			/**
+			 * Sets the {@link ReactiveSessionRegistry} to use.
+			 * @param reactiveSessionRegistry the {@link ReactiveSessionRegistry} to use
+			 * @return the {@link ConcurrentSessionsSpec} to continue customizing
+			 */
+			public ConcurrentSessionsSpec sessionRegistry(ReactiveSessionRegistry reactiveSessionRegistry) {
+				SessionManagementSpec.this.sessionRegistry = reactiveSessionRegistry;
+				return this;
+			}
+
+			/**
+			 * Sets the {@link Function} to use to determine the maximum number of
```

Comment:
It's not clear to me how I'd state that for some authentications, there is no maximum. For example, is it:

```java
.maxSessions((authentication) -> containsSomeAuthority(authentication) ? Mono.just(Integer.MAX_VALUE) : Mono.just(3))
```

or something else?

(I can see in the reference that it is -1, but this may be value to place in the JavaDoc, too. Also, since it is a magic number, you might consider a static constant like `UNLIMITED`. This adds clarity to the user's intent and it also gives you the ability to change that value in the future should it become necessary.)

Reply:
Instead of the user dealing with magic numbers, I think it would be better to define a `SessionLimit` POJO with a `SessionLimit#of(int)` and `SessionLimit#unlimited()` static factory methods to achieve a better declarative configuration.

In addition to that, there are now `maximumSessions(SessionLimit)` and `maximumSessions(Function<Authentication, Mono<SessionLimit>)` available.